src/BuiltInTools/DotNetDeltaApplier/HotReloadAgent.cs

```diff
@@ -45,116 +44,211 @@ private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
             }
         }
 
-        private sealed class UpdateHandlerActions
+        internal sealed class UpdateHandlerActions
+        {
+            public List<Action<Type[]?>> Before { get; } = new();
+            public List<Action<Type[]?>> After { get; } = new();
+            public List<Action<Type[]?>> ClearCache { get; } = new();
+            public List<Action<Type[]?>> UpdateApplication { get; } = new();
+        }
+
+        private UpdateHandlerActions GetMetadataUpdateHandlerActions()
         {
-            public UpdateHandlerActions(List<Action<Type[]?>> before, List<Action<Type[]?>> after)
+            // In a typical app, a handful of assemblies will have metadata handler attributes. We need to discover these assemblies
+            // and ensure that they are topologically sorted so that handlers in a dependency are executed before the dependent (e.g.
+            // the reflection cache action in System.Private.CoreLib is executed before System.Text.Json clears it's own cache.)
+            // This would ensure that there is a well-defined order and that caches and updates more lower in the application stack
+            // are up to date before ones higher in the stack are recomputed.
+            var assemblies = GetAssembliesWithMetadataUpdateHandlerAttributes();
+            var sortedAssemblies = TopologicalSort(CollectionsMarshal.AsSpan(assemblies));
```

Comment:
There is no point in using `CollectionsMarshal.AsSpan` here. Change it to use a regular enumerator instead. 

The potentially dangerous APIs like ``CollectionsMarshal.AsSpan` are meant to be only used for core that needs to be micro-optimized to the last cycle. It is not the case here. You are calling GetCustomAttributes inside the loop that is like 1000x times more expensive than the foreach.